packages/angular/src/wallet/wallet.errors.ts

```diff
@@ -0,0 +1,29 @@
+import { WalletError } from '@solana/wallet-adapter-base';
+
+export class WalletNotSelectedError extends WalletError {
+    constructor() {
+        super();
+        this.name = 'WalletNotSelectedError';
+    }
+}
+
+export class SignTransactionNotFoundError extends WalletError {
+    constructor() {
+        super();
+        this.name = 'SignTransactionNotFoundError';
+    }
+}
+
+export class SignAllTransactionsNotFoundError extends WalletError {
+    constructor() {
+        super();
+        this.name = 'SignAllTransactionsNotFoundError';
+    }
+}
+
+export class SignMessageNotFoundError extends WalletError {
+    constructor() {
+        super();
+        this.name = 'SignMessageNotFoundError';
+    }
+}
```

Comment:
It would be ideal if there was a way to not throw runtime errors when these are used, which would mean this package wouldn't need these errors. The way the React package for this works is that feature detection is done in a TypeScript compatible way, so if the method is defined by the adapter, it's callable. Is there a way we can do that here?

Reply:
Let's see if I get this right. We want the `signMessage` to be of type `((message: Uint8Array) => Observable<Uint8Array>) | undefined`?

From my understanding, that means consumers would have to do something like:

```ts
if (signMessage) {
    signMessage(message)
}
```

And that goes for all the sign-methods, right? If that's the case it seems like a doable thing.

Aside from that, in RxJs, if an error is thrown inside a stream like I'm doing there it won't terminate the **app**, instead, it would terminate the **stream**. This way consumers can have a handler for the error to inspect if something went wrong, but if they don't add the handler, the app won't break because the error is not treated as an unhandled exception at the application level.

Reply:
Yeah, exactly. In the React package, if they do `const { signTransaction } = useWallet()` the function can be undefined, so they'll get a tsc error if they try to use it without checking. This also encourages dapp developers to not rely on all wallets having all features, since using the common `sendTransaction` interface is the golden path

Reply:
I hear ya.

I'll try to figure that out, given that the store is a class and the sign-methods are instance methods, it might not be trivial.

On the other hand, the sign-methods I use atm return an observable, in order for the observable to fire, the consumer needs to call .subscribe, if my method returns undefined they would get a compilation error, forcing them to check before subscribing.

I think that if the sign-methods have a type of `(input) => Observable | undefined`, it will behave as we expect, the function definitions won't be undefined but at least the library won't allow consumers to subscribe without checking. I know this isn't exactly analogous but would have the behavior intended.

Reply:
Just for clarity, for example if the `signMessage` is of type `(message: Uint8Array) => Observable<Uint8Array> | undefined`, a consumer trying to sign a message would have to:

```typescript
const signMessage$ = signMessage(message);

if (!signMessage$) {
     // notify error and return
}

signMessage$.subscribe()
```

The lazy nature of an observable makes it that just calling the function doesn't trigger `adapter.<method>`, but we have to subscribe. If the "factory method" creating the observable returns undefined calling `.subscribe` will fail at compilation.

Reply:
Huh, interesting. Yeah, I think as long as it's reasonably idiomatic/expected behavior for Angular/Rxjs users, and will produce TS errors instead of runtime, that's fine.

Reply:
IMHO, the error is what I'd expect, but the main goal is to make it as similar to the react package as possible. If at some point in time it's decided to switch to an error in React, I'll gladly update this. 

I'll have that fixed by next week and will include the rest of the feedback, currently buidling something cool for Ignition so I'm juggling with a lot atm.

Reply:
The core issue as I see it is that dapps can't be expected handle different wallets differently, and they need a way to feature detect. Otherwise the error has to be thrown and handled at runtime, or bubble up to the user, which is bad for everyone because they aren't able to do anything about it.

Take your time and good luck in the hackathon :)

Reply:
I totally get the reasoning behind it and I think it's a cool design, is just that I find errors a bit more expressive to know what's going on, as a dev I'm used to try/catch every promise/observable as a rule of thumb. Don't get wrong, I'm a bit Angular-biased, and since I use a lot of classes and Object-oriented stuff, having a "conditional method" makes some noise to me. 

Thanks a lot man!!